src/java/org/apache/cassandra/transport/Server.java

```diff
@@ -306,6 +315,24 @@ int countConnectedClients()
             return allChannels.size() != 0 ? allChannels.size() - 1 : 0;
         }
 
+        int countConnectedClients(Predicate<ServerConnection> predicate)
+        {
+            int count = 0;
+            for (Channel c : allChannels)
```

Comment:
what happens when connection is closed while you iterate over the channels? what if channels are closed?

Reply:
I'd expect that even if the Channel is closed, the attributes will still be present. If the `ServerConnection` does happen to be null, the instanceof check will fail.  The `ClientState` on the `connection` (which the predicates evaluate) will never be null, so it should generally be safe.

Also, if an exception is thrown here, the calling thread will be a JMX rmi connection thread, which will simply bubble the exception up to the client, so I think it will behave ok even in exceptional states.

Reply:
:+1: let me think about how I can test this.  The challenge will be that I think to reproduce it the channel will need to be closed.  This approach was borrowed from other methods on the class (`countConnectedClientsByUser`) so maybe there is some existing test precedence I can pull from.

Alternatively I could probably inject an unexpected failure using byteman..will keep thinking on this.

Reply:
@tolbertam you could also probably mock that when `c.attr(Connection.attributeKey).get()` is called that get() returns null or throws.

Reply:
Added `ConnectionTrackerTest` that includes:

* Validate that `Channel` lacking a `Connection.attributeKey` get skipped over.
* Validate that `Channel` having a `Connection.attributeKey` that is not a `ServerConnection` get skipped.
* Validate that if `c.attr(Connection.attributeKey).get()` throws an unhandled exception it gets propagated up to the caller (ultimately preventing the calculation from working, which I think is the behavior we want, but let me know if you think that should be different).
* Also added tests that validate the expected basic functionality of `countConnectedClients(predicate)` and `countConnectedClientsByUser` (which I didn't change in this PR but figured would be nice to have at a unit level).

Reply:
great! that is enough I guess ...